### DISCUSSION TITLE
RFC: 致谢环境

### DISCUSSION THREAD

**Discussion by yetiansh** [stone-zeng/fduthesis#253] [Ideas]
希望有一个单独的致谢环境以和摘要环境的目录以及正文风格保持一致（以及满足强迫症）。用chapter等方法实现致谢，想与摘要align比较麻烦。

目前自己按照摘要环境加了个致谢环境，由于完全不了解dtx编写，遇到两个问题。

按照[这里](https://github.com/stone-zeng/fduthesis/blob/a1bf5281f2289a28cb6cb5bee1f0c79ab0981801/source/fduthesis.dtx#L5454-L5481)的代码：
```latex
% \begin{environment}{abstract}
% \begin{environment}{abstract*}
% \changes{v0.7}{2018/03/05}{整理代码。}
% 摘要环境。在中文模板定义了中英文双语摘要，但在英文模板中则没有
% 定义中文摘要。
%    \begin{macrocode}
\NewDocumentEnvironment { abstract  } { }
%<class>  { \@@_abstract_begin:    } { \@@_abstract_end:      }
%<class-en>  { \@@_abstract_en_begin: } { \@@_abstract_en_end:   }
%<*class>
\NewDocumentEnvironment { abstract* } { }
  { \@@_abstract_en_begin: } { \@@_abstract_en_end:   }
%</class>
%    \end{macrocode}
% \end{environment}
% \end{environment}
%
% \begin{macro}{\@@_abstract_begin:,\@@_abstract_en_begin:}
% 摘要页标题。
%    \begin{macrocode}
%<*class>
\cs_new_protected:Npn \@@_abstract_begin:
  { \@@_chapter:V \c_@@_name_abstract_tl    }
%</class>
\cs_new_protected:Npn \@@_abstract_en_begin:
  { \@@_chapter:V \c_@@_name_abstract_en_tl }
%    \end{macrocode}
% \end{macro}
```

我将abstract全部替换成acknowledgement，将acknowledgement_end置空，即：
```latex
% \begin{macro}{\@@_acknowledgement_end:,\@@_acknowledgement_en_end:}
%    \begin{macrocode}
%<*class>
\cs_new_protected:Npn \@@_acknowledgement_end:
  {
  }
%</class>
\cs_new_protected:Npn \@@_acknowledgement_en_end:
  {
  }
%    \end{macrocode}
% \end{macro}
```

Q1：编译后遇到如下error:
```
) (./sources/acknowledgement.tex
! Undefined control sequence.
<argument> \c__fdu_name_acknowledgement_tl 
                                           
l.2 \begin{acknowledgement}
                           
? x
```
显然与`\c_@@_name_abstract_tl`相同，`\c_@@_name_acknowledgement_tl`需要被替换成"致谢"，但未找到`\c_@@_name_abstract_tl`是如何被替换成"摘要"的？

Q2：将将acknowledgement_end置空是否已经足够？

**Comment by atxy-blip** [comment]
> 未找到 `\c_@@_name_abstract_tl` 是如何被替换成"摘要"的？

在 https://github.com/stone-zeng/fduthesis/blob/a1bf5281f2289a28cb6cb5bee1f0c79ab0981801/source/fduthesis.dtx#L6301 通过 `\@@_define_name:nnn` 定义。

> 将acknowledgement_end置空是否已经足够？

足够了。用上述方法创建的致谢环境实际是个无编号的 chapter。

**Reply by yetiansh** [reply]
感谢，我已经改过了

**Comment by stone-zeng** [comment]
致谢环境已经添加了，见 d8f86e3。

**Reply by MusicJIA** [reply]
您好，致谢是这样使用的吗？
\begin{acknowledgement}
致谢
\end{acknowledgement}。

我用了“https://github.com/stone-zeng/fduthesis/commit/d8f86e3398ad218dc6fb9b057a5c8eec4142b5d1” 中的fduthesis-doc.dtx

**Reply by sikouhjw** [reply]
是

**Reply by SwitWu** [reply]
加个 s，`acknowledgements`

**Reply by MusicJIA** [reply]
编译时出现的问题：Environment acknowledgements undefined. \begin(acknowledgements}。
但是，我已经用了您给的链接中的“fduthesis-doc.dtx” 这个文件

**Reply by SwitWu** [reply]
你用 dtx 干什么，你需要用的是 `fduthesis.cls`

**Reply by MusicJIA** [reply]
替换了我之前的fduthesis.cls，编译还是出现相同的问题@@##。

**Reply by SwitWu** [reply]
反正 Releases v0.9a 是有的

**Reply by MusicJIA** [reply]
嗯嗯。谢谢啦！
